cpp/ql/src/experimental/Security/CWE/CWE-561/FindIncorrectlyUsedSwitch.ql

```diff
@@ -0,0 +1,137 @@
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-561
+ *       external/cwe/cwe-691
+ *       external/cwe/cwe-478
+ */
+
+import cpp
+import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+import semmle.code.cpp.commons.Exclusions
+
+/** Holds if the range contains no boundary values. */
+predicate isRealRange(Expr exp) {
+  upperBound(exp).toString() != "18446744073709551616" and
+  upperBound(exp).toString() != "9223372036854775807" and
+  upperBound(exp).toString() != "4294967295" and
+  upperBound(exp).toString() != "Infinity" and
+  upperBound(exp).toString() != "NaN" and
+  lowerBound(exp).toString() != "-9223372036854775808" and
+  lowerBound(exp).toString() != "-4294967296" and
+  lowerBound(exp).toString() != "-Infinity" and
+  lowerBound(exp).toString() != "NaN" and
+  upperBound(exp) != 2147483647 and
+  upperBound(exp) != 268435455 and
+  upperBound(exp) != 33554431 and
+  upperBound(exp) != 8388607 and
+  upperBound(exp) != 65535 and
+  upperBound(exp) != 32767 and
+  upperBound(exp) != 255 and
+  upperBound(exp) != 127 and
+  lowerBound(exp) != -2147483648 and
+  lowerBound(exp) != -268435456 and
+  lowerBound(exp) != -33554432 and
+  lowerBound(exp) != -8388608 and
+  lowerBound(exp) != -65536 and
+  lowerBound(exp) != -32768 and
+  lowerBound(exp) != -128
+  or
+  lowerBound(exp) = 0 and
+  upperBound(exp) = 1
+}
+
+/** Holds if the range of values for the condition is less than the choices. */
+predicate isNotAllSelected(SwitchStmt swtmp) {
+  not swtmp.getExpr().isConstant() and
+  exists(int i |
+    i != 0 and
+    (
+      i = lowerBound(swtmp.getASwitchCase().getExpr()) and
+      upperBound(swtmp.getExpr()) < i
+      or
+      (
+        i = upperBound(swtmp.getASwitchCase().getExpr()) or
+        i = upperBound(swtmp.getASwitchCase().getEndExpr())
+      ) and
+      lowerBound(swtmp.getExpr()) > i
+    )
+  )
+}
+
+/** Holds if the range of values for the condition is greater than the selection. */
+predicate isConditionBig(SwitchStmt swtmp) {
+  not swtmp.hasDefaultCase() and
+  not exists(int iu, int il |
+    (
+      iu = upperBound(swtmp.getASwitchCase().getExpr()) or
+      iu = upperBound(swtmp.getASwitchCase().getEndExpr())
+    ) and
+    upperBound(swtmp.getExpr()) = iu and
+    (
+      il = lowerBound(swtmp.getASwitchCase().getExpr()) or
+      il = lowerBound(swtmp.getASwitchCase().getEndExpr())
+    ) and
+    lowerBound(swtmp.getExpr()) = il
+  )
+}
+
+/** Holds if there are labels inside the block with names similar to `default` or `case`. */
+predicate isWrongLableName(SwitchStmt swtmp) {
+  not swtmp.hasDefaultCase() and
+  exists(LabelStmt lb |
+    (
+      (
+        lb.getName().charAt(0) = "d" or
+        lb.getName().charAt(0) = "c"
+      ) and
+      (
+        lb.getName().charAt(1) = "e" or
+        lb.getName().charAt(1) = "a"
+      ) and
+      (
+        lb.getName().charAt(2) = "f" or
+        lb.getName().charAt(2) = "s"
+      )
```

Comment:
This will match labels with the names "def", "des", "daf", "das", "cef", "ces", "caf", "cas". Is that the intended effect?

Would it be better to flag _all_ `LabelStmt` inside switch statements that have no `GotoStmt` jumping to them?

Reply:
> This will match labels with the names "def", "des", "daf", "das", "cef", "ces", "caf", "cas". Is that the intended effect?

I was hoping to find all the labels that start with these characters.

 
> Would it be better to flag _all_ `LabelStmt` inside switch statements that have no `GotoStmt` jumping to them?

your offer is much better from a security point of view. it will also cover situations when you forgot to write `case`. I'll look at it live and change the code.

Reply:
Good day.

my analysis shows that unused labels are the result of constructs using complex transitions such as

https://lgtm.com/projects/g/gperftools/gperftools/snapshot/3a9817e8738c18d5ef40f444ec79c18cacc365b8/files/src/tests/stacktrace_unittest.cc?sort=name&dir=ASC&mode=heatmap#xa2c70e59f3daf555:1

or a macro that generates various labels

https://lgtm.com/projects/g/fluent/fluent-bit/snapshot/9c2dbdae7d7779fa364e2298c101e4510498e18d/files/lib/onigmo/regexec.c?sort=name&dir=ASC&mode=heatmap#xc7d8130fce95515:1

or the result of `#ifdef` constraints

https://lgtm.com/projects/g/SoftEtherVPN/SoftEtherVPN/snapshot/39fd835476c762d03fba693e9ee7f7bb3a275787/files/src/Cedar/Client.c?sort=name&dir=ASC&mode=heatmap#x24cfa9cdeb8f67e:1


unfortunately in all these situations, the developer assumes the presence of unmanaged labels.

I want to find a label similar to `default` for example `default` `dafalt` `defalt` `defaul` and so on.

which is possible and is used in the transition operator (or maybe not),
but it creates the possibility that one of the developers might mistake this label as part of a `switch` construct.
therefore, such a label is bad, at least from a stylistic point of view.

if you do not like the current code, I suggest changing it to detect specific erroneous labels,
 with similar ones on `default` it is not difficult to do this.
with labels similar to `case`, I'll think about it, because I wanted to take into account, among other things, `caseTARARATARARA`

Reply:
> my analysis shows that unused labels are the result of constructs using complex transitions such as
> 
> https://lgtm.com/projects/g/gperftools/gperftools/snapshot/3a9817e8738c18d5ef40f444ec79c18cacc365b8/files/src/tests/stacktrace_unittest.cc?sort=name&dir=ASC&mode=heatmap#xa2c70e59f3daf555:1

To filter out such labels you could require that there doesn't exist a [`LabelLiteral`](https://github.com/github/codeql/blob/565af1a879b59a253351c7f4d68cfb9e525a9dad/cpp/ql/src/semmle/code/cpp/exprs/Literal.qll#L40) whose `getLabel` gives you back the label.

------

I admit those labels are all quite... intricate... I don't particularly mind this solution. Have you found any bugs caused by a misspelled `default:`?

Reply:
at the initial stage of the formation of the request, I found two ambiguous situations of using labels. but this did not look like an example of an error.
I keep looking.